analyzers/src/SonarAnalyzer.CSharp/Rules/TimestampsShouldNotBeUsedAsPrimaryKeys.cs

```diff
@@ -0,0 +1,106 @@
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace SonarAnalyzer.Rules.CSharp;
+
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public sealed class TimestampsShouldNotBeUsedAsPrimaryKeys : SonarDiagnosticAnalyzer
+{
+    private const string DiagnosticId = "S3363";
+    private const string MessageFormat = "'{0}' should not be used as primary key";
+
+    private static readonly KnownType[] TemporalTypes = new[]
+    {
+        KnownType.System_DateOnly,
+        KnownType.System_DateTime,
+        KnownType.System_DateTimeOffset,
+        KnownType.System_TimeSpan,
+        KnownType.System_TimeOnly
+    };
+    private static readonly string[] KeyAttributeTypeNames = TypeNamesForAttribute(KnownType.System_ComponentModel_DataAnnotations_KeyAttribute);
+    private static readonly DiagnosticDescriptor Rule = DescriptorFactory.Create(DiagnosticId, MessageFormat);
+
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+    protected override void Initialize(SonarAnalysisContext context) =>
+        context.RegisterCompilationStartAction(start =>
+        {
+            if (!IsReferencingEntityFramework(start.Compilation))
+            {
+                return;
+            }
+
+            context.RegisterNodeAction(c =>
+            {
+                // To improve performance the attributes and property types are only matched by their names, rather than their actual symbol.
+                // This results in a couple of FNs, but those scenarios are very rare.
+                var classDeclaration = (ClassDeclarationSyntax)c.Node;
+                var className = classDeclaration.Identifier.ValueText;
+                var keyProperties = classDeclaration.Members
+                    .OfType<PropertyDeclarationSyntax>()
+                    .Where(x => IsPublicReadWriteProperty(x) && IsTemporalType(x) && IsKeyProperty(x, className));
+
+                foreach (var keyProperty in keyProperties)
+                {
+                    c.ReportIssue(Diagnostic.Create(Rule, keyProperty.Type.GetLocation(), keyProperty.Type.GetName()));
+                }
+            }, SyntaxKind.ClassDeclaration);
+        });
+
+    private static bool IsReferencingEntityFramework(Compilation compilation) =>
+        compilation.GetTypeByMetadataName(KnownType.Microsoft_EntityFrameworkCore_DbContext) is not null
+        || compilation.GetTypeByMetadataName(KnownType.Microsoft_EntityFramework_DbContext) is not null;
+
+    private static bool IsPublicReadWriteProperty(PropertyDeclarationSyntax property) =>
+        property.Modifiers.Any(x => x.IsKind(SyntaxKind.PublicKeyword))
+        && property.AccessorList is not null
+        && property.AccessorList.Accessors.Any(x => x.Keyword.IsKind(SyntaxKind.GetKeyword))
```

Comment:
Ou can use IsAnyKind(SyntaxKind.GetKeyword, SyntaxKind.SetKeyword)

Reply:
I'm looking for a read-write property, as Entity Framework won't map read-only/calculated properties to database keys. So I need to check if it has both a getter and a setter, not any of them.